File: sdk/zorynex.ts
/**
 * Zorynex TypeScript SDK
 * ========================
 * Single-file client. Copy into your project — no npm install needed.
 * Works in Node.js (18+), Deno, Bun, and modern browsers.
 *
 * Usage:
 *   import { ZorynexClient } from './zorynex';
 *
 *   const client = new ZorynexClient({
 *     baseUrl:  'http://127.0.0.1:8000',
 *     apiKey:   'dev-key',
 *     tenantId: 'default',
 *   });
 *
 *   // Simple: governance auto-resolved
 *   const proof = await client.recordDecision({
 *     instanceId: 'loan-001',
 *     fromState:  'pending',
 *     toState:    'approved',
 *     rawInputs:  { credit_score: '742' },
 *   });
 *   console.log(proof.proofId);
 */

// ── Types ─────────────────────────────────────────────────────────────────────

export interface ZorynexConfig {
  baseUrl:  string;
  apiKey:   string;
  tenantId: string;
  timeout?: number;   // ms, default 30000
}

export interface DecisionRequest {
  instanceId:    string;
  fromState:     string;
  toState:       string;
  rawInputs?:    Record<string, string>;
  // Optional — auto-resolved from approved lists if omitted
  modelVersion?:  string;
  agentVersion?:  string;
  policyVersion?: string;
  reasonCode?:    string;
  policyRule?:    string;
  featureContributions?: Array<{ feature: string; contribution: string }>;
  thresholdUsed?: string;
  metadata?:      Record<string, string>;
}

export interface DecisionResponse {
  proofId:     string;
  sequenceId:  number;
  instanceId:  string;
  currentHash: string;
  proofUrl:    string;
  traceId:     string;
}

export interface VerifyPackageResponse {
  verified:      boolean;
  instanceId:    string | null;
  finalState:    string | null;
  proofCount:    number;
  modelVersion:  string | null;
  policyVersion: string | null;
  signingKey:    string | null;
  checks: Array<{
    name:    string;
    passed:  boolean;
    detail:  string | null;
    failure: string | null;
  }>;
  traceId: string;
}

export class ZorynexError extends Error {
  constructor(
    public readonly statusCode: number,
    public readonly detail: unknown,
  ) {
    super(`HTTP ${statusCode}: ${JSON.stringify(detail)}`);
    this.name = 'ZorynexError';
  }
}

// ── Client ────────────────────────────────────────────────────────────────────

export class ZorynexClient {
  private readonly base:    string;
  private readonly headers: Record<string, string>;
  private readonly timeout: number;

  constructor(config: ZorynexConfig) {
    this.base    = config.baseUrl.replace(/\/$/, '');
    this.timeout = config.timeout ?? 30_000;
    this.headers = {
      'X-API-Key':    config.apiKey,
      'X-Tenant-Id':  config.tenantId,
      'Content-Type': 'application/json',
    };
  }

  // ── HTTP helpers ────────────────────────────────────────────────────────────

  private async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);

    try {
      const resp = await fetch(`${this.base}${path}`, {
        method,
        headers: this.headers,
        body:    body !== undefined ? JSON.stringify(body) : undefined,
        signal:  controller.signal,
      });

      const data = await resp.json();
      if (!resp.ok) throw new ZorynexError(resp.status, data);
      return data as T;
    } finally {
      clearTimeout(timer);
    }
  }

  private get<T>(path: string) { return this.request<T>('GET', path); }
  private post<T>(path: string, body: unknown) { return this.request<T>('POST', path, body); }

  // ── Quickstart ──────────────────────────────────────────────────────────────

  /** Bootstrap a complete demo environment in one call. */
  bootstrap() {
    return this.post<Record<string, unknown>>('/demo/bootstrap', {});
  }

  // ── Governance ──────────────────────────────────────────────────────────────

  approveModel(name: string, version: string) {
    return this.post('/governance/model', { name, version });
  }

  approveAgent(name: string, version: string) {
    return this.post('/governance/agent', { name, version });
  }

  approvePolicy(name: string, version: string) {
    return this.post('/governance/policy', { name, version });
  }

  governanceStatus() {
    return this.get('/governance/status');
  }

  // ── Protocol + instance ─────────────────────────────────────────────────────

  compileProtocol(params: {
    states:       string[];
    initialState: string;
    transitions?: Array<{ from_state: string; to_state: string }>;
    metadata?:    Record<string, string>;
  }) {
    return this.post<{ protocol_hash: string }>('/protocol/compile', {
      states:        params.states,
      initial_state: params.initialState,
      transitions:   params.transitions ?? [],
      metadata:      params.metadata    ?? {},
    });
  }

  createInstance(instanceId: string, protocolHash?: string) {
    return this.post<{ instance_id: string; initial_state: string }>(
      '/instance/create',
      { instance_id: instanceId, ...(protocolHash ? { protocol_hash: protocolHash } : {}) },
    );
  }

  // ── Decisions ───────────────────────────────────────────────────────────────

  /**
   * Record an AI decision as a cryptographic proof.
   *
   * Minimal usage (governance auto-resolved):
   *   await client.recordDecision({
   *     instanceId: 'loan-001', fromState: 'pending', toState: 'approved',
   *     rawInputs: { credit_score: '742' }
   *   });
   */
  recordDecision(req: DecisionRequest): Promise<DecisionResponse> {
    const body: Record<string, unknown> = {
      instance_id: req.instanceId,
      from_state:  req.fromState,
      to_state:    req.toState,
      raw_inputs:  req.rawInputs ?? {},
    };
    if (req.modelVersion)          body['model_version']          = req.modelVersion;
    if (req.agentVersion)          body['agent_version']          = req.agentVersion;
    if (req.policyVersion)         body['policy_version']         = req.policyVersion;
    if (req.reasonCode)            body['reason_code']            = req.reasonCode;
    if (req.policyRule)            body['policy_rule']            = req.policyRule;
    if (req.featureContributions)  body['feature_contributions']  = req.featureContributions;
    if (req.thresholdUsed)         body['threshold_used']         = req.thresholdUsed;
    if (req.metadata)              body['metadata']               = req.metadata;

    return this.post<DecisionResponse>('/decision', body);
  }

  // ── Proofs ──────────────────────────────────────────────────────────────────

  getProof(instanceId: string, opts: { sequenceId?: number; verbose?: boolean } = {}) {
    const params = new URLSearchParams();
    if (opts.sequenceId) params.set('sequence_id', String(opts.sequenceId));
    if (opts.verbose)    params.set('verbose', 'true');
    const qs = params.toString();
    return this.get(`/proof/${instanceId}${qs ? '?' + qs : ''}`);
  }

  getChain(instanceId: string, full = false) {
    return this.get(`/chain/${instanceId}${full ? '?full=true' : ''}`);
  }

  exportProof(instanceId: string, inline = true) {
    return this.get(`/proof/export/${instanceId}${inline ? '?inline=true' : ''}`);
  }

  // ── Verification ────────────────────────────────────────────────────────────

  verifyProof(proofDict: Record<string, unknown>) {
    return this.post('/verify', proofDict);
  }

  verifyPackage(pkg: Record<string, unknown>): Promise<VerifyPackageResponse> {
    return this.post<VerifyPackageResponse>('/verify-package', pkg);
  }

  // ── Audit ────────────────────────────────────────────────────────────────────

  chainVerify() { return this.get('/audit/chain-verify'); }
  compliancePack() { return this.get('/audit/compliance'); }

  // ── Health ────────────────────────────────────────────────────────────────────

  health() { return this.get('/health'); }
  ready()  { return this.get('/ready'); }
}

File: sdk/zorynex.py
"""
Zorynex Python SDK
====================
Single-file client. No external dependencies beyond the standard library.
Copy this file into your project and import it — no pip install needed.

Usage:
    from zorynex import ZorynexClient

    client = ZorynexClient(
        base_url = "http://127.0.0.1:8000",
        api_key  = "dev-key",
        tenant_id= "default",
    )

    # Quick: let the platform auto-resolve governance
    proof = client.record_decision(
        instance_id = "loan-001",
        from_state  = "pending",
        to_state    = "approved",
        raw_inputs  = {"credit_score": "742"},
    )
    print(proof["proof_id"])

    # Full: explicit governance
    proof = client.record_decision(
        instance_id    = "loan-001",
        from_state     = "pending",
        to_state       = "approved",
        model_version  = "credit-model-v3.1",
        agent_version  = "underwriter-v1.0",
        policy_version = "credit-policy-v2",
        reason_code    = "SCORE_ABOVE_THRESHOLD",
        raw_inputs     = {"credit_score": "742"},
    )
"""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any


class ZorynexError(Exception):
    def __init__(self, status_code: int, detail: Any) -> None:
        self.status_code = status_code
        self.detail      = detail
        super().__init__(f"HTTP {status_code}: {detail}")


class ZorynexClient:
    """
    Thin HTTP client for the Zorynex Provable AI API.

    All methods return the parsed JSON response as a dict.
    Raises ZorynexError on HTTP 4xx/5xx.
    """

    def __init__(
        self,
        base_url:  str = "http://127.0.0.1:8000",
        api_key:   str = "dev-key",
        tenant_id: str = "default",
        timeout:   float = 30.0,
    ) -> None:
        self._base    = base_url.rstrip("/")
        self._headers = {
            "X-API-Key":    api_key,
            "X-Tenant-Id":  tenant_id,
            "Content-Type": "application/json",
        }
        self._timeout = timeout

    # ── HTTP helpers ──────────────────────────────────────────────────────────

    def _request(self, method: str, path: str, body: dict | None = None) -> dict:
        url  = self._base + path
        data = json.dumps(body).encode() if body is not None else None
        req  = urllib.request.Request(url, data=data, headers=self._headers, method=method)
        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                return json.loads(resp.read())
        except urllib.error.HTTPError as e:
            try:
                detail = json.loads(e.read())
            except Exception:
                detail = e.reason
            raise ZorynexError(e.code, detail)

    def _get(self, path: str) -> dict:
        return self._request("GET", path)

    def _post(self, path: str, body: dict) -> dict:
        return self._request("POST", path, body)

    # ── Quickstart ────────────────────────────────────────────────────────────

    def bootstrap(self) -> dict:
        """
        Seed a complete demo environment in one call.
        Approves a model/agent/policy, compiles a protocol, creates an instance.
        Returns the instance_id and next-step payload.
        """
        return self._post("/demo/bootstrap", {})

    # ── Governance ────────────────────────────────────────────────────────────

    def approve_model(self, name: str, version: str) -> dict:
        """Approve a model version for use in decisions."""
        return self._post("/governance/model", {"name": name, "version": version})

    def approve_agent(self, name: str, version: str) -> dict:
        """Approve an agent version for use in decisions."""
        return self._post("/governance/agent", {"name": name, "version": version})

    def approve_policy(self, name: str, version: str) -> dict:
        """Approve a policy version for use in decisions."""
        return self._post("/governance/policy", {"name": name, "version": version})

    def governance_status(self) -> dict:
        """Return all approved models, agents, and policies."""
        return self._get("/governance/status")

    # ── Protocol + instance ───────────────────────────────────────────────────

    def compile_protocol(
        self,
        states:        list[str],
        initial_state: str,
        transitions:   list[dict] | None = None,
        metadata:      dict | None = None,
    ) -> dict:
        """Define a workflow protocol. Returns {protocol_hash}."""
        return self._post("/protocol/compile", {
            "states":        states,
            "initial_state": initial_state,
            "transitions":   transitions or [],
            "metadata":      metadata    or {},
        })

    def create_instance(
        self,
        instance_id:   str,
        protocol_hash: str | None = None,
    ) -> dict:
        """Create a workflow instance. Returns {instance_id, initial_state}."""
        body: dict = {"instance_id": instance_id}
        if protocol_hash:
            body["protocol_hash"] = protocol_hash
        return self._post("/instance/create", body)

    # ── Decisions ─────────────────────────────────────────────────────────────

    def record_decision(
        self,
        instance_id:    str,
        from_state:     str,
        to_state:       str,
        raw_inputs:     dict | None = None,
        # Optional — auto-resolved if omitted
        model_version:  str | None = None,
        agent_version:  str | None = None,
        policy_version: str | None = None,
        reason_code:    str | None = None,
        policy_rule:    str | None = None,
        # Optional extras
        feature_contributions: list[dict] | None = None,
        threshold_used:        str | None = None,
        metadata:              dict | None = None,
    ) -> dict:
        """
        Record an AI decision as a cryptographic proof.

        Minimal usage (governance auto-resolved):
            client.record_decision("loan-001", "pending", "approved",
                                   raw_inputs={"credit_score": "742"})

        Full usage (all fields explicit):
            client.record_decision("loan-001", "pending", "approved",
                                   model_version="credit-model-v3.1",
                                   agent_version="underwriter-v1.0",
                                   policy_version="credit-policy-v2",
                                   reason_code="SCORE_ABOVE_THRESHOLD",
                                   raw_inputs={"credit_score": "742"})

        Returns: {proof_id, sequence_id, instance_id, current_hash, proof_url}
        """
        body: dict = {
            "instance_id": instance_id,
            "from_state":  from_state,
            "to_state":    to_state,
            "raw_inputs":  raw_inputs or {},
        }
        if model_version:            body["model_version"]  = model_version
        if agent_version:            body["agent_version"]  = agent_version
        if policy_version:           body["policy_version"] = policy_version
        if reason_code:              body["reason_code"]    = reason_code
        if policy_rule:              body["policy_rule"]    = policy_rule
        if feature_contributions:    body["feature_contributions"] = feature_contributions
        if threshold_used:           body["threshold_used"] = threshold_used
        if metadata:                 body["metadata"]       = metadata

        return self._post("/decision", body)

    # ── Proofs ────────────────────────────────────────────────────────────────

    def get_proof(self, instance_id: str, sequence_id: int | None = None,
                  verbose: bool = False) -> dict:
        """Get a proof by instance_id. Add verbose=True for full canonical payload."""
        path = f"/proof/{instance_id}"
        params = []
        if sequence_id: params.append(f"sequence_id={sequence_id}")
        if verbose:     params.append("verbose=true")
        if params: path += "?" + "&".join(params)
        return self._get(path)

    def get_chain(self, instance_id: str, full: bool = False) -> dict:
        """Get chain summary. Add full=True for all entries."""
        path = f"/chain/{instance_id}"
        if full: path += "?full=true"
        return self._get(path)

    def export_proof(self, instance_id: str, inline: bool = True) -> dict:
        """Export a complete verifiable proof package."""
        path = f"/proof/export/{instance_id}"
        if inline: path += "?inline=true"
        return self._get(path)

    # ── Verification ──────────────────────────────────────────────────────────

    def verify_proof(self, proof_dict: dict) -> dict:
        """Verify a single proof dict. Returns {valid, ...}."""
        return self._post("/verify", proof_dict)

    def verify_package(self, package: dict) -> dict:
        """
        Verify a complete exported proof package.
        Returns {verified, checks, instance_id, final_state, ...}.
        """
        return self._post("/verify-package", package)

    # ── Audit ─────────────────────────────────────────────────────────────────

    def chain_verify(self) -> dict:
        """Verify the integrity of the full proof ledger for this tenant."""
        return self._get("/audit/chain-verify")

    def compliance_pack(self) -> dict:
        """Export the compliance JSON (SR 11-7 / EU AI Act / CFPB)."""
        return self._get("/audit/compliance")

    # ── Health ────────────────────────────────────────────────────────────────

    def health(self) -> dict:
        """Check if the server is alive. No auth required."""
        return self._request("GET", "/health")

    def ready(self) -> dict:
        """Check if the server is ready (DB + signer available). No auth required."""
        return self._request("GET", "/ready")